pinecone/client.go

```diff
@@ -463,16 +476,75 @@ func toCollection(cm *control.CollectionModel) *Collection {
 	}
 }
 
-func decodeCollection(resBody io.ReadCloser) (*Collection, error) {
+func decodeCollection(resBodyBytes []byte) (*Collection, error) {
 	var collectionModel control.CollectionModel
-	err := json.NewDecoder(resBody).Decode(&collectionModel)
+	err := json.Unmarshal(resBodyBytes, &collectionModel)
 	if err != nil {
 		return nil, fmt.Errorf("failed to decode collection response: %w", err)
 	}
 
 	return toCollection(&collectionModel), nil
 }
 
+func decodeErrorResponse(resBodyBytes []byte) (*control.ErrorResponse, error) {
+	var errorResponse control.ErrorResponse
+	err := json.Unmarshal(resBodyBytes, &errorResponse)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode error response: %w", err)
+	}
+
+	if errorResponse.Status == 0 {
+		return nil, fmt.Errorf("unable to parse ErrorResponse: %v", string(resBodyBytes))
+	}
+
+	return &errorResponse, nil
+}
+
+type errorResponseMap struct {
+	StatusCode int    `json:"status_code"`
+	Body       string `json:"body,omitempty"`
+	ErrorCode  string `json:"error_code,omitempty"`
+	Message    string `json:"message,omitempty"`
+	Details    string `json:"details,omitempty"`
+}
+
+func handleErrorResponseBody(resBodyBytes []byte, statusCode int, errMsgPrefix string) error {
```

Comment:
When testing I noticed that we don't always seem to be guaranteed the `control.ErrorResponse` shape for the body, which would lead to specific errors when trying to decode things.

What I tried to do here was attempt to decode into `ErrorResponse`, and if that fails we just parse the whole body out as a string.